### ISSUE TITLE
Error when running several DE genes per cluster

### ISSUE THREAD

**Issue by ccruizm** [markziemann/mitch#12]
Good day,

I want to try your tool. I have tried to apply it to my dataset (scRNA-10x preprocessed by Seurat). When I run the functions for one cluster it (`y <- mitch_import(markers_0, DEtype="Seurat")`) it seems to work but once I make a list of DE genes per cluster and try to run it, I get an error:
```
markers_0 <- FindMarkers(seu_obj, ident.1 = 0, min.pct = 0.25)
markers_1 <- FindMarkers(seu_obj, ident.1 = 1, min.pct = 0.25)
markers_2 <- FindMarkers(seu_obj, ident.1 = 2, min.pct = 0.25)
markers_3 <- FindMarkers(seu_obj, ident.1 = 3, min.pct = 0.25)
markers_4 <- FindMarkers(seu_obj, ident.1 = 4, min.pct = 0.25)

x <- list('clus_0'=markers_0,'clus_1'=markers_1,'clus_2'=markers_2,'clus_3'=markers_3,
         'clus_4'=markers_4)
y <- mitch_import(x, DEtype="Seurat")

Warning message in mitch_rank(input_profile):
“Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.”
Warning message in mitch_rank(input_profile):
“Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.”
Warning message in mitch_rank(input_profile):
“Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.”
Warning message in mitch_rank(input_profile):
“Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.”
Warning message in mclapply(X, FUN, ..., mc.cores = mc.cores, mc.preschedule = mc.preschedule, :
“all scheduled cores encountered errors in user code”
Error in as.data.frame.default(x[[i]], optional = TRUE, stringsAsFactors = stringsAsFactors): cannot coerce class ‘"try-error"’ to a data.frame
Traceback:

1. mitch_calc(y, genesets, priority = "significance", cores = 4)
2. MANOVA(ranked_profile, genesets, minsetsize = minsetsize, cores = cores, 
 .     priority = priority)
3. ldply(res, data.frame)
4. llply(.data = .data, .fun = .fun, ..., .progress = .progress, 
 .     .inform = .inform, .parallel = .parallel, .paropts = .paropts)
5. structure(lapply(pieces, .fun, ...), dim = dim(pieces))
6. lapply(pieces, .fun, ...)
7. FUN(X[[i]], ...)
8. as.data.frame(x[[i]], optional = TRUE, stringsAsFactors = stringsAsFactors)
9. as.data.frame.default(x[[i]], optional = TRUE, stringsAsFactors = stringsAsFactors)
10. stop(gettextf("cannot coerce class %s to a data.frame", sQuote(deparse(class(x))[1L])), 
  .     domain = NA)
```

The data.frame `y`  contains multiple NA per column since not all DE genes per cluster are the same. What would be the best way to overcome this issue?

**Comment by markziemann** [comment]
Hi @ccruizm  there could be different reasons for this. 
Could you show the result of 
```
head(markers_0)
head(y)
head(genesets)
```
I just want to see whether the gene names in the profiling data match the gene sets you have

**Comment by ccruizm** [comment]
Sure, thanks for the quick reply

<img width="706" alt="Screenshot 2020-07-01 at 17 10 45" src="https://user-images.githubusercontent.com/37718031/86260618-cf56b280-bbbd-11ea-9dac-bfc1a79e0d44.png">

**Comment by markziemann** [comment]
Thanks for this. I suspect it may have something to with the NA values. Tomorrow I will run the Seurat immune alignment (https://satijalab.org/seurat/v3.1/immune_alignment.html) workflow and see whether this error is reproducible

**Comment by markziemann** [comment]
I've completed the workflow as shown here: https://github.com/markziemann/mitch_example_workflows/blob/master/mitch_for_seurat.Rmd

It seems to have worked properly, but with the one concern is that some of the p-values are NA which might be related to your problem

`
> head(res$enrichment_result,10)
                                                                             set
854                                       TRAF3-dependent IRF activation pathway
239                                            Eukaryotic Translation Elongation
571                                                     Peptide chain elongation
387                                              Interferon alpha/beta signaling
752                                                     Selenocysteine synthesis
525 Nonsense Mediated Decay (NMD) independent of the Exon Junction Complex (EJC)
241                                           Eukaryotic Translation Termination
751                                                  Selenoamino acid metabolism
934                                                       Viral mRNA Translation
479                                      Modulation by Mtb of host immune system
    setSize      pMANOVA      s.pDC        s.Mk        s.DC s.CD14mono
854       6 0.9955782983  0.2718491  0.19795557  0.36788040  0.3447354
239      82 0.0001113113 -0.3999491 -0.10943189 -0.34347845 -0.2992342
571      80 0.0001113113 -0.3989031 -0.10943189 -0.34242796 -0.2993088
387      46 0.3466835950  0.3764816  0.17447939  0.32609217  0.2945829
752      79           NA -0.3954039 -0.08818241 -0.34053367 -0.2964281
525      81           NA -0.3989762 -0.08818241 -0.33639079 -0.2953200
241      81           NA -0.3976032 -0.08818241 -0.33438562 -0.2942433
751      81           NA -0.3954039 -0.05784426 -0.33793455 -0.2954665
934      80           NA -0.3887833 -0.08818241 -0.33324284 -0.2986596
479       5           NA  0.4464993  0.14039945 -0.01897482 -0.3064481
    s.CD16mono     s.Bact        s.B      s.CD8T       s.NK      s.Tact
854  0.2856783  0.1551029  0.1548543  0.10278821  0.1630679  0.17491956
239 -0.2318515 -0.1671393 -0.1679304 -0.12608257 -0.1686028 -0.04514053
571 -0.2338313 -0.1671393 -0.1679304 -0.12608257 -0.1693899 -0.04514053
387  0.2378079  0.1595366  0.1489591  0.09897456  0.1617409  0.15911892
752 -0.2331913 -0.1651883 -0.1659078 -0.12510531 -0.1696343  0.02041731
525 -0.2325792 -0.1662540 -0.1662595 -0.12510531 -0.1694839  0.02041731
241 -0.2331913 -0.1651883 -0.1659078 -0.12510531 -0.1696343  0.02041731
751 -0.2337853 -0.1651883 -0.1659078 -0.12510531 -0.1696343  0.02041731
934 -0.2331913 -0.1651883 -0.1659078 -0.12510531 -0.1696343  0.02041731
479 -0.2531841  0.1753610  0.1802757  0.12239413  0.1774582  0.04470688
    s.CD4naiveT   s.CD4memT        p.pDC         p.Mk         p.DC   p.CD14mono
854  0.08332778  0.07736011 6.128354e-01 4.960027e-01 4.721234e-01 3.761230e-01
239 -0.10140455 -0.09958884 1.283906e-06 7.439924e-06 6.535406e-07 1.338327e-05
571 -0.10140455 -0.09958884 1.283906e-06 7.439924e-06 6.535406e-07 1.338327e-05
387  0.08438089  0.07195798 2.303920e-02 4.673302e-02 3.589888e-02 6.003482e-02
752 -0.09891981 -0.09857763           NA           NA           NA           NA
525 -0.09949800 -0.09857763           NA           NA           NA           NA
241 -0.09891981 -0.09857763           NA           NA           NA           NA
751 -0.09891981 -0.09857763           NA           NA           NA           NA
934 -0.09891981 -0.09857763           NA           NA           NA           NA
479  0.10696457  0.09889377           NA           NA           NA           NA
      p.CD16mono       p.Bact          p.B       p.CD8T         p.NK
854 3.425996e-01 5.283011e-01 5.950835e-01 5.176462e-01 5.566109e-01
239 2.951394e-06 1.731493e-06 4.799701e-06 2.244431e-05 2.235415e-06
571 2.951394e-06 1.731493e-06 4.799701e-06 2.244431e-05 2.235415e-06
387 4.377532e-02 7.040607e-03 1.881754e-02 1.422735e-02 4.886027e-02
752           NA           NA           NA           NA           NA
525           NA           NA           NA           NA           NA
241           NA           NA           NA           NA           NA
751           NA           NA           NA           NA           NA
934           NA           NA           NA           NA           NA
479           NA           NA           NA           NA           NA
          p.Tact  p.CD4naiveT    p.CD4memT    s.dist         SD p.adjustMANOVA
854 4.589045e-01 5.339551e-01 5.294298e-01 0.7599981 0.09805120   0.9999444025
239 1.756804e-05 8.361923e-05 7.119515e-05 0.7455915 0.10885177   0.0009906704
571 1.756804e-05 8.361923e-05 7.119515e-05 0.7453745 0.10859921   0.0009906704
387 4.647302e-02 2.377028e-02 2.223460e-02 0.7369281 0.09746025   0.4392148037
752           NA           NA           NA 0.7358604 0.11828079             NA
525           NA           NA           NA 0.7356065 0.11819869             NA
241           NA           NA           NA 0.7333449 0.11770513             NA
751           NA           NA           NA 0.7314405 0.12033558             NA
934           NA           NA           NA 0.7298729 0.11650380             NA
479           NA           NA           NA 0.7144151 0.20016354             NA
`

I think that the error in the MANOVA() function could stem from contrasts in which there are no detected genes. Could you try to run the analysis with a smaller number of contrasts and see whether it persists?

**Comment by ccruizm** [comment]
Sorry for the late reply. I could also run the tutorial but my data still get the same error. If I run the functions only with one cluster `y <- mitch_import(markers_0, DEtype="Seurat")` and `res <- mitch_calc(y, genesets, priority="significance", cores = 4)`, I get the results but the charts and report do not look like anything compared with the tutorial. I could share a downsampled Seurat object to see whether you can reproduce the problem. Thanks in advance!

**Comment by markziemann** [comment]
Hi Cristian, if you could provide the downsampled Seurat object, that would be very helpful to resolve the issue.

**Comment by ccruizm** [comment]
Great. You can download the file here https://drive.google.com/file/d/1mnk4shzkrjAOe0VJfE7P9kV5MQUywQRm/view?usp=sharing
Hopefully, you will find where the issue is. Thanks!

**Comment by markziemann** [comment]
Thanks for sharing. 
I ran the analysis and got this far
```
> res <- mitch_calc(y, genesets, priority = "significance", cores = 4)
  |================================================================================| 100%, Elapsed 00:01
Note: When prioritising by significance (ie: small 
            p-values), large effect sizes might be missed.
Warning messages:
1: In mitch_rank(input_profile) :
  Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.
2: In mitch_rank(input_profile) :
  Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.
3: In mitch_rank(input_profile) :
  Warning: >60% of genes have the same score. This isn't
            optimal for rank based enrichment analysis.
> head(res$enrichment_result)
                                                  set setSize      pMANOVA
93                              Developmental Biology     157 5.069381e-06
109          Epigenetic regulation of gene expression      14 5.069381e-06
142                   Gene expression (Transcription)     145 5.069381e-06
143                     Generic Transcription Pathway     110 5.069381e-06
249 Positive epigenetic regulation of rRNA expression      11 5.069381e-06
263                   RNA Polymerase II Transcription     130 5.069381e-06
        s.clus_0    s.clus_1    s.clus_2    s.clus_3    s.clus_4  p.clus_0
93  -0.059079046  0.13815204 -0.06378076 -0.05967931  0.04924609 0.9132922
109  0.104697087 -0.03272508  0.08899049 -0.03094615 -0.24164554 0.9132922
142  0.074607330 -0.03026916  0.09510341 -0.05272786 -0.15674934 0.9132922
143 -0.005699422 -0.02631551  0.07577494 -0.03619524 -0.07280095 0.9132922
249  0.110618653 -0.01013265  0.08899049 -0.03094615 -0.20544515 0.9132922
263  0.048236205 -0.03083849  0.08288358 -0.03814643 -0.12107702 0.9132922
        p.clus_1   p.clus_2 p.clus_3  p.clus_4    s.dist         SD
93  0.0002071529 0.04397595 0.357888 0.4001543 0.1806408 0.09031387
109 0.0002071529 0.04397595 0.357888 0.4001543 0.2816060 0.13857285
142 0.0002071529 0.04397595 0.357888 0.4001543 0.2070694 0.10234344
143 0.0002071529 0.04397595 0.357888 0.4001543 0.1143543 0.05528506
249 0.0002071529 0.04397595 0.357888 0.4001543 0.2518409 0.12548271
263 0.0002071529 0.04397595 0.357888 0.4001543 0.1620562 0.07994903
    p.adjustMANOVA
93    2.872649e-05
109   2.872649e-05
142   2.872649e-05
143   2.872649e-05
249   2.872649e-05
263   2.872649e-05
```
I'm using R4.0 with mitch v1.0.8

But then I did have an error when it came to rendering the charts, specifically for the contour plots in ggpairs, probably because in some panels there are no genes detected. I will need to make some modifications.

**Comment by ccruizm** [comment]
I updated mitch from 1.0.6 to 1.0.8 but still the same problem. I am also using R 4.0.0. I do not understand where the problem is. Did you run the same code I shared in the jupyter notebook?

**Comment by markziemann** [comment]
Thanks for this, here is what I got http://118.138.234.73/public/blog/run_mitch.html

The only difference between mine and yours is  that you analysed markers 0-6 and I did 0-4.

Could you try with 0-4 only?

Next, if you send me the "y" object made from markers 0-6 I hopefully should be able to reproduce the error.

**Comment by ccruizm** [comment]
Indeed I could run it when using markers from 0 to 4, but once I included them all, the error shows up again. What do you think is the cause of this strange behavior? Below the 'y' object with all the clusters.
I also have a problem with the report. Hopefully, you can figure it out 
Thanks!
[mitch.zip](https://github.com/markziemann/mitch/files/5346464/mitch.zip)

**Comment by markziemann** [comment]
Cristian, thanks for sending this. I was able to figure out the origin of the problem.
Basically the data is too sparse to work properly with the R MANOVA commands.

```
fit <- manova(x ~ inset)
sumAOV <- summary.aov(fit)
 Error in summary.manova(fit) : residuals have rank 8 < 9
```
It yields an error because of the lack of degrees of freedom, which as I understand is due to having more contrasts than gene measurements. 

Although in this case MANOVA may not work, it should be possible to perform the individual ANOVAs and aggregate the results. Unfortunately this is not a quick fix and will require some things to be reworked. 

For your analysis I would suggest limiting the number of contrasts to 4 or 5. The approach below worked well for me where I selected contrasts with 400 or more detected genes.

```
yy <- y[,which(apply(y,2,function(x) sum(as.integer(is.finite(x)))   ) > 400)]
res <- mitch_calc(yy, genesets, priority="significance")
```

If you're finding issues with mitch_plots or mitch_reports, try installing mitch_1.1.10. Currently only available by install_github() but will be pushed to BioC soon.
